src/org/apache/hadoop/tools/Curl.php

```diff
@@ -194,13 +196,30 @@ private function _exec($options, $returnInfo = false)
         }
         if (isset($this->options['local_file_handler'])) {
             $fp = $this->options['local_file_handler'];
-            flock($fp,LOCK_EX);
+            flock($fp, LOCK_EX);
             $options[CURLOPT_WRITEFUNCTION] = function ($ch, $string) use ($fp) {
                 $length = fwrite($fp, $string);
+
                 return $length;
             };
             fflush($fp);
         }
+        // auto add content-length header
+        $has_content_length_header = false;
+        foreach ($options[CURLOPT_HTTPHEADER] as $header) {
+            if (stripos($header, 'content-length:') === 0) {
+                $has_content_length_header = true;
+                break;
+            }
+        }
+        if (!$has_content_length_header && !isset($options[CURLOPT_INFILE]) && !isset($options[CURLOPT_INFILESIZE])) {
+            $length = 0;
+            if (isset($options[CURLOPT_POSTFIELDS])) {
+                $length = strlen($options[CURLOPT_POSTFIELDS]);
```

Comment:
actually one more possible caveat; it looks like `CURLOPT_POSTFIELDS` could be a `@` prefixed string which implies a file name that is then read in (deprecated in PHP 5.5 though) or it could even be a key/value array, on which `strlen` will fail. To be safe better also check if `is_string` and that the string does not begin with `@`.

Reply:
Oh yes, this library is >= php5.4.0 ... will make that change now